packages/coreutils/src/url.ts

```diff
@@ -58,9 +58,12 @@ export namespace URLExt {
     const slashes = colon && header![3];
 
     // Construct the URL prefix.
-    const prefix = shorthand
-      ? '//'
-      : [protocol, colon, slashes].filter(str => str).join('');
+    let prefix = [protocol, colon, slashes].filter(str => str).join('');
```

Comment:
How about we fix the header regex above to correctly recognize the protocol instead of make the mistake? According to the RFC, the [scheme](https://tools.ietf.org/html/rfc3986#section-3.1) can only be something like `ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )`. Can we fix the `top.match` line above to correctly parse this?

Another useful resource is the matching ABNF for relative paths: https://tools.ietf.org/html/rfc3986#section-3.3

Reply:
Another thing we can do is use the built-in browser URL object to try to parse things. That does all the correct work for us, and is probably much safer, etc: https://developer.mozilla.org/en-US/docs/Web/API/URL

Reply:
Using the browser functionality is probably sound, but I can't remember if we needed `coreutils` to work in `node`.

The trouble is that we aren't dealing with URI/URL values, so it is sound to have the scheme shorthand `//www.example.com/foo/bar` which tells the browser to use whichever protocol is already being used.

Reply:
Of course we'd have to use the base parameter for `URL` if we had a relative url. This works fine, for example: `new URL('//a.b.c/d/e', 'https://myurl.com/')`.

Reply:
In fact, I think we can just always put in the base parameter for URL no matter what, and the browser will use it if it should.

Now, working in node - right, that is another issue. I think it's probably still good to correct our url parsing above to conform to the RFC.

Reply:
huh, except node has a URL package too: https://nodejs.org/api/url.html

Reply:
Right, there are two issues, I agree with you that using `URL(...)` is better all around.

It looks like at some point we meticulously made sure to check if `document` exists before doing browser-specific things, but right now the `time.ts` module appears to refer to `document.documentElement` without checking.

I think we wanted `coreutils` to be useful in a CLI environment, but maybe the time-format functions have never needed to be used in `node` so we didn't notice. Although since node ships with a URL lib, maybe we can conditionally import it if it doesn't exist the way Lumino does with `window.crypto` vs. `requires('crypto')` ...

Reply:
This conversation was too async, we arrived at the same conclusions :joy:

Reply:
`let jasonResponsd = await dariansConclusion;` :)

Reply:
Ok so I'm having trouble following the async convo, what is the conclusion exactly?

Reply:
I think the idea is that we should switch the implementation of this function from parsing the strings ourselves to using the `URL` object but also making sure it works both in `node` and in the browser.